maint: pandas 2.0 forward compatible changes

jjerphan opened the pull request:
#### Reference Issues/PRs

Fixes #237.
Follow-up of #343.

#### What does this implement/fix? Explain your changes.

This PR brings support for pandas 2.0. by:
 - removing bound in environment specifications
 - adapting the normalization and the denormalization of `pandas.Series` and `pandas.DataFrame`
 - adapting tests for changes of behavior due to the new defaults of pandas 2.0 

#### Any other comments?

We need to discuss how to properly support some changes of behavior of pandas with respect to:
 - [the specification of the resolution for the `datetime64` or `timedelta64`](https://pandas.pydata.org/docs/dev/whatsnew/v2.0.0.html#construction-with-datetime64-or-timedelta64-dtype-with-unsupported-resolution)
 - [`datetime64` and `timedelta64` conversion for their resolution](https://pandas.pydata.org/docs/dev/whatsnew/v2.0.0.html#disallow-astype-conversion-to-non-supported-datetime64-timedelta64-dtypes) introduced by https://github.com/pandas-dev/pandas/issues/48928
 
Edges cases which I think can be treated subsequently:
 - https://github.com/man-group/ArcticDB/pull/540/files#r1265126115
 - https://github.com/man-group/ArcticDB/pull/540/files#r1288498933
 - https://github.com/man-group/ArcticDB/pull/540/files#r1273260829
 - https://github.com/man-group/ArcticDB/pull/540/files#r1288531578
 
See `NOTE` and `TODO` in the diff of this PR.

jjerphan commented on python/tests/integration/arcticdb/version_store/test_categorical.py:
```diff
@@ -109,6 +138,25 @@ def test_categorical_with_integers_and_strings(lmdb_version_store, sym):
     # should be category
     assert read_df.cat_int.dtype == "category"
     assert read_df.cat_str.dtype == "category"
+    # TODO: understand why Windows with Pandas 2.0 fails with this error:
+    # AssertionError: Attributes of DataFrame.iloc[:, 1] (column name="cat_int") are different
+    # Attribute "dtype" are different
+    # [left]:  CategoricalDtype(categories=[0, 1, 2, 3, 4, 5], ordered=False)
+    # [right]: CategoricalDtype(categories=[0, 1, 2, 3, 4, 5], ordered=False)
```
```suggestion
```

jjerphan commented on python/tests/integration/arcticdb/version_store/test_categorical.py:
```diff
@@ -81,19 +84,45 @@ def test_categorical_empty(lmdb_version_store, sym):
     # should be category
-    assert read_df.cat.dtype == "category"
+    assert read_df.cat_int.dtype == "category"
+    # TODO: understand why Windows with Pandas 2.0 fails with this error:
+    # AssertionError: Attributes of DataFrame.iloc[:, 1] (column name="cat_int") are different
+    # Attribute "dtype" are different
+    # [left]:  CategoricalDtype(categories=[0, 1, 2, 3, 4, 5], ordered=False)
+    # [right]: CategoricalDtype(categories=[0, 1, 2, 3, 4, 5], ordered=False)
```
```suggestion
```

mehertz commented on python/arcticdb/util/_versions.py:
```diff
@@ -0,0 +1,7 @@
+import pandas as pd
```
BSL header please

qc00 commented on python/arcticdb/version_store/_normalization.py:
```diff
@@ -178,12 +187,19 @@ def _to_primitive(arr, arr_name, dynamic_strings, string_max_len=None, coerce_co
+                # See: https://github.com/pandas-dev/pandas/issues/17261
+                # We want to maintain consistent behaviour, so we treat empty series as containing floats.
+                # val_type = ValueType::FLOAT;
+                coerce_column_type = float
```
We have to consider #224 first.

The concern is if the user did not intend for the column to be `float` but happened to have no data for the initial write, then it wouldn't be possible to append to this symbol subsequently.

jjerphan replied:
Is there a way to change the type of an (empty) column?

jjerphan replied:
I would wait for https://github.com/man-group/ArcticDB/pull/646/ to be merged before this PR first. What do you think?

willdealtry replied:
Can you just get rid of coerce_column_type entirely, and all associated tests, and we'll just record a NoneType now that 646 is merged?

jjerphan replied:
Done in https://github.com/man-group/ArcticDB/pull/804.

jjerphan commented on python/tests/unit/arcticdb/version_store/test_filtering.py:
```diff
@@ -27,7 +27,7 @@
 from arcticdb_ext.exceptions import InternalException, UserInputException
 from arcticdb.util.test import assert_frame_equal
-from arcticdb.util._versions import IS_PANDAS_ZERO
+from arcticdb.util._versions import PANDAS_VERSION, IS_PANDAS_ZERO
```
Indeed. This is fixed by f751693.

poodlewars commented:
We discussed this today. We should merge as much of this PR as we can but retain the pin. We don't need to wait till NoneType support.

Afterwards we can add NoneType support and have the Pandas 2.0 implementation make use of that. At that point we can remove the Pandas pin.

jjerphan commented:
Though slightly in favour for waiting (is there an urgency to merge changes without support for pandas 2?), I am OK for merging now with the pin.

If we merge now, we need to change the title of this PR.

qc00 commented on python/arcticdb/version_store/_normalization.py:
```diff
@@ -497,6 +514,20 @@ def _index_to_records(self, df, pd_norm, dynamic_strings, string_max_len):
         else:
+            n_rows = len(index)
+            n_categorical_columns = len(df.select_dtypes(include="category").columns)
+            if IS_PANDAS_TWO and isinstance(index, (Index, RangeIndex)) and n_rows == 0 and n_categorical_columns == 0:
```
Both `RangeIndex` and `DatetimeIndex` are sub-types of `Index`, so the `isinstance(index, (Index, RangeIndex))` part is probably always True.

qc00 commented on python/arcticdb/version_store/_normalization.py:
```diff
@@ -497,6 +514,20 @@ def _index_to_records(self, df, pd_norm, dynamic_strings, string_max_len):
+                # columns.
+                #
+                # The index is converted to a DatetimeIndex for preserving the behavior of ArcticDB with Pandas 1.0.
+                index = DatetimeIndex([])
```
I came up with two tests for safe normalisation:

* The data frame we read back should be equivalent to the input
* If the user had a particular (index) type in mind but accidentally generated a zero-row data frame initially, the user should be able to append to such a frame

So the best solution is to introduce something in the C++ to allow the index type of an empty DF:

* to be accurately saved
* to change on the first append

qc00 commented on python/arcticdb/version_store/_normalization.py:
```diff
@@ -660,7 +698,10 @@ def denormalize(self, item, norm_meta):
+            # Specifying the column dtype since pandas might just convert the type of empty columns to something
+            # else if we don't specify explicitly cast it depending on the version.
+            columns_dtype = None if data is None else {name: np_array.dtype for name, np_array in data.items()}
+            df = DataFrame(data, index=index, columns=columns).astype(dtype=columns_dtype)
```
Use the `dtype` parameter of the `DataFrame` constructor. Otherwise, `astype` does a very expensive copy.

jjerphan replied:
IIRC, I tried, but the `dtype` keyword does not respect the specification and it is converting them to something else. This horrible solution is the first one I have found to respect the specification of dtypes.

qc00 commented on python/tests/integration/arcticdb/version_store/test_basic_version_store.py:
```diff
@@ -1725,7 +1726,11 @@ def test_dynamic_schema_column_hash_update(lmdb_version_store_column_buckets):
-    assert_frame_equal(vit.data.astype("float"), df)
+    if not IS_PANDAS_TWO:
+        # Pandas 1.0 does not cast to float natively, but Pandas 2.0 does.
+        vit.data = vit.data.astype("float")
```
Is there any harm in doing this in Pandas 2 as well?

I don't want to give any impression that there's Pandas version-dependent behaviour in this test

jjerphan replied:
In Pandas < 2.0, updating a `DataFrame` uniquely storing integers with another `DataFrame` that is uniquely storing integers changes the dtypes to float64 (yes).

This is fixed in Pandas >= 2.0.

The comment is not accurate, and I will update it.

jjerphan replied:
Updated in 97928e1.

qc00 commented on python/tests/integration/arcticdb/version_store/test_categorical.py:
```diff
@@ -81,19 +84,40 @@ def test_categorical_empty(lmdb_version_store, sym):
     # should be category
-    assert read_df.cat.dtype == "category"
+    assert read_df.cat_int.dtype == "category"
+    if IS_PANDAS_TWO and (sys.maxsize <= 2**32 or sys.platform.startswith("win32")):
```
We do not support 32-bit platforms anyway?

jjerphan replied:
Removed in 5a698e2.

qc00 commented on python/tests/stress/arcticdb/version_store/test_stress_multicolumn.py:
```diff
@@ -67,6 +69,10 @@ def test_stress_multicolumn(lib_type, request):
             [pd.DataFrame([dict_merge(dict(date=date), sd) for sd in sec_data]) for date in p], ignore_index=True
         )
 
+        if IS_PANDAS_TWO:
```
An ideal normalisation should not require any special handling in unrelated tests. This is a symptom of breaking my first test.

jjerphan replied:
This has been removed in 5565ef1.

jjerphan commented:
I've added the pin on pandas again and changed the title of this PR as discussed.

The last comment of this PR can be addressed in a subsequent one after #646 is merged.

qc00 commented on python/arcticdb/version_store/_normalization.py:
```diff
@@ -660,7 +698,15 @@ def denormalize(self, item, norm_meta):
+            # some (empty) columns to another one and since the `dtype` keyword for `pd.DataFrame` constructor
+            # does not accept a mapping such as `columns_dtype`...
+            for column_name, dtype in columns_dtype.items():
+                df[column_name] = df[column_name].astype(dtype, copy=False)
```
An idea just occurred to me: can you try converting `columns` to a sequence of `pandas.Series` first and then construct the DF with `copy=False`?

jjerphan replied:
I tried this:

```diff
diff --git i/python/arcticdb/version_store/_normalization.py w/python/arcticdb/version_store/_normalization.py
index 280bbaf6..4fab43ba 100644
--- i/python/arcticdb/version_store/_normalization.py
+++ w/python/arcticdb/version_store/_normalization.py
@@ -698,14 +698,8 @@ class DataFrameNormalizer(_PandasNormalizer):
         columns, denormed_columns, data = _denormalize_columns(item, norm_meta, idx_type, n_indexes)
 
         if not self._skip_df_consolidation:
-            columns_dtype = {} if data is None else {name: np_array.dtype for name, np_array in data.items()}
-            df = DataFrame(data, index=index, columns=columns)
-
-            # Horribly setting the columns' dtype manually, since pandas might just convert the dtype of
-            # some (empty) columns to another one and since the `dtype` keyword for `pd.DataFrame` constructor
-            # does not accept a mapping such as `columns_dtype`...
-            for column_name, dtype in columns_dtype.items():
-                df[column_name] = df[column_name].astype(dtype, copy=False)
+            columns_mapping = {} if data is None else {name: pd.Series(np_array, index=index, dtype=np_array.dtype) for name, np_array in data.items()}
+            df = DataFrame(index=index, columns=columns_mapping, copy=False)
 
         else:
             if index is not None:

```

but this creates a `pd.DataFrame` full of NaNs whilst Series are sane. :shrug:

willdealtry replied:
I think that _skip_df_normalization is always true, and we want it to be, as the alternative is slow: https://github.com/man-group/ArcticDB/blob/002fa24473cbf222825c65fb6c08c6d8904b1669/python/arcticdb/version_store/library.py#L292

jjerphan replied:
Might the consolidation not actually be always skipped due to this definition?

https://github.com/man-group/ArcticDB/blob/6f5b9146f3e1f72ae40bbda7bbcca491c3a81a91/python/arcticdb/version_store/_normalization.py#L662

willdealtry replied:
Yes. I think we should remove this entirely at some later point, as we always want to skip it

jjerphan commented on python/tests/unit/arcticdb/version_store/test_empty_writes.py:
```diff
@@ -134,12 +135,45 @@ def test_empty_series(lmdb_version_store_dynamic_schema, sym):
     lmdb_version_store.write(sym, df)
-    assert lmdb_version_store.is_symbol_pickled(sym)
-    assert_frame_equal(df, lmdb_version_store.read(sym).data)
+
```
I guess we might want to wait for https://github.com/man-group/ArcticDB/pull/646 to be merged first.

willdealtry replied:
Just kill this test, it's testing something that we aren't actually doing anything any more

willdealtry commented on python/arcticdb/version_store/_normalization.py:
```diff
@@ -169,6 +170,14 @@ def _to_primitive(arr, arr_name, dynamic_strings, string_max_len=None, coerce_co
+        # See: https://pandas.pydata.org/docs/dev/whatsnew/v2.0.0.html#construction-with-datetime64-or-timedelta64-dtype-with-unsupported-resolution  # noqa: E501
+        # We want to maintain consistent behaviour, so we convert any other resolution
+        # to `datetime64[ns]`.
+        arr = arr.astype(DTN64_DTYPE, copy=False)
```
I think that's a good idea, but out of scope for this PR

willdealtry replied:
Can we for the time being just change this to copy=True?

jjerphan replied:
Fixed by https://github.com/man-group/ArcticDB/pull/801.

jjerphan commented:
FYI, I have added references to the remaining threads there in https://github.com/man-group/ArcticDB/issues/634#issuecomment-1687810744.

As discussed with Alex and William, let's merge this PR and treat them in dedicated PRs.
